design-proposals/vm-delete-protection.md

```diff
@@ -0,0 +1,147 @@
+for backwards compatibility details.
+
+Pros:
+
+- Kubernetes built-in feature which is straightforward to maintain, implement and configure.
+- ValidationAdmisssionPolicy is [GA](https://kubernetes.io/blog/2024/03/12/kubernetes-1-30-upcoming-changes/) since Kuberentes version 1.30.
+
+Cons:
+
+- It will not work in deployments using Kubernetes versions \< 1.30
+
+A proof of concept implementation can be found
+at: [https://github.com/kubevirt/kubevirt/commit/7167e68723201b3899f8f36b02e5bf11701ee722](https://github.com/kubevirt/kubevirt/commit/7167e68723201b3899f8f36b02e5bf11701ee722)
+
+#### Option 2\#: Virtual Machine Admission Webhook
+
+A new validation rule is added to the VM admitters. The admitter will check for presence of
+the \`[kubevirt.io/vm-delete-protection](http://kubevirt.io/vm-delete-protection)\` annotation, filter delete
+operations, and will reject any attempt to delete VM objects if the annotation is present and its value is set to “true”
+or “True”.
+
+Pros:
+
+- It works with any Kubernetes version.
+
+Cons:
+
+- It will increase the maintenance effort.
+
+A proof of concept implementation can be found
+at:  [https://github.com/kubevirt/kubevirt/commit/2fafcf649019c26c8802838f2bbe673a9e93f04c](https://github.com/kubevirt/kubevirt/commit/2fafcf649019c26c8802838f2bbe673a9e93f04c)
+
+#### Comparison between Option \#1 and Option \#2
+
+We have carried out some benchmark experiments to determine if there is any impact in terms of performance when the user
+deletes VMs. In the test, we created a bunch of VMs and delete them at the same time using the command:
+
+```bash
+$ kubectl delete –all 
+```
+
+We have defined three sets of experiments changing the number of VMs created, and the running state of the VM when it is
+deleted, i.e., stopped or running. The values shown here are the mean of 31 runs:
+
+| Option     | Deletion time stopped 400 VMs(s) | Deletion time stopped 40 VMs(s) | Deletion time running 40 VMs(s) |
+|:-----------|:---------------------------------|:--------------------------------|:--------------------------------|
+| Baseline   | 78.97                            | 6.03                            | 198.52                          |
+| Option \#1 | 78.96                            | 6.23                            | 199.61                          |
+| Option \#2 | 79.06                            | 6.16                            | 199.8                           |
+
+Please note that these tests are not aimed to compare exact deletion times between different experiment sets, but to
+compare them to a baseline and check if any option implies a degradation and if there is a faster option than the other.
+Differences between options are negligible in any scenario analyzed. Moreover, **no performance degradation has been
+observed.**
+
+Given the pros and cons, **OPTION \#1 is preferred in this case.**
```

Comment:
Will we go with this option then?

Reply:
IMHO, yes, we should go with this option.